aten/src/ATen/SparseTensorImpl.h

```diff
@@ -1,10 +1,83 @@
 #pragma once
 
+#include "ATen/Tensor.h"
 #include "ATen/TensorImpl.h"
+#include "ATen/Error.h"
 
 namespace at {
 struct SparseTensorImpl : public TensorImpl {
-  explicit SparseTensorImpl(Type * type)
-  : TensorImpl(type) {}
+  // Stored in COO format, indices + values
+
+  std::vector<int64_t> size_;
+  // INVARIANT: indices_.size(1) >= nnz_ && values_.size(0) >= nnz_
+  int64_t nnz_ = 0;
+  int64_t dimI_ = 0; // number of sparse dimensions
+  int64_t dimV_ = 0; // number of dense dimensions
+
+  // 2-D tensor of nDim x nnz of indices. May have nnz dim bigger than nnz
+  // as buffer, so we keep track of both
+  Tensor indices_; // always a LongTensor
+  Tensor values_;
+
+  // A sparse tensor is 'coalesced' if every index occurs at most once in
+  // the indices tensor, and the indices are in sorted order.
+  // Most math operations can only be performed on ordered sparse tensors
```

Comment:
what does ordered mean here?  do you just mean coalesced?

Reply:
yep, coalesced